File: src/empathy_ai.py
"""
Empathetic AI Module
Implements AI with empathy, context awareness, and ethical responses
Supports multiple models for different scenarios
"""
from typing import Dict, List, Optional, Tuple
from conversation_history import ConversationHistory
from safety_filter import SafetyFilter
import json
import random


class EmpathyAI:
    """Main AI class with empathy, memory, and safety features"""
    
    # Model profiles for different contexts
    MODELS = {
        "empathetic": {
            "name": "Empathetic Model",
            "specialization": "emotional support and understanding",
            "traits": ["compassionate", "validating", "patient", "non-judgmental"],
            "best_for": "crisis support, emotional topics, personal struggles"
        },
        "practical": {
            "name": "Practical Model",
            "specialization": "actionable advice and solutions",
            "traits": ["efficient", "focused", "solution-oriented", "clear"],
            "best_for": "problem-solving, advice, technical questions"
        },
        "analytical": {
            "name": "Analytical Model",
            "specialization": "deep thinking and complex topics",
            "traits": ["thoughtful", "thorough", "balanced", "nuanced"],
            "best_for": "philosophy, ethics, complex analysis"
        },
        "supportive": {
            "name": "Supportive Model",
            "specialization": "encouragement and motivation",
            "traits": ["positive", "encouraging", "motivating", "uplifting"],
            "best_for": "motivation, goal-setting, self-improvement"
        }
    }
    
    # Empathy patterns for responses
    EMPATHY_STARTERS = [
        "I understand how you feel.",
        "That sounds really challenging.",
        "Your feelings make complete sense.",
        "I hear you, and what you're going through matters.",
        "That's a difficult situation, and I appreciate you sharing it.",
        "I can see why that would be frustrating/painful/concerning.",
    ]
    
    def __init__(self, user_name: str = "User", default_model: str = "empathetic"):
        """
        Initialize EmpathyAI
        
        Args:
            user_name: Name of the user for personalization
            default_model: Default model to use
        """
        self.user_name = user_name
        self.current_model = default_model
        self.history = ConversationHistory()
        self.safety_filter = SafetyFilter(sensitivity_level="medium")
        self.emotional_context = {}
        
    def select_best_model(self, message: str) -> str:
        """
        Select the best model based on message content
        
        Args:
            message: User message to analyze
            
        Returns:
            Selected model name
        """
        message_lower = message.lower()
        
        # Detect crisis or emotional distress
        emotional_keywords = ["feel", "sad", "lonely", "depressed", "anxious", 
                            "scared", "worried", "hurt", "pain", "suffering"]
        if any(keyword in message_lower for keyword in emotional_keywords):
            return "empathetic"
        
        # Detect problem-solving requests
        problem_keywords = ["how to", "help me", "advice", "solution", "fix", "problem"]
        if any(keyword in message_lower for keyword in problem_keywords):
            return "practical"
        
        # Detect philosophical/complex topics
        complex_keywords = ["why", "meaning", "ethics", "philosophy", "right or wrong", "moral"]
        if any(keyword in message_lower for keyword in complex_keywords):
            return "analytical"
        
        # Detect motivation/goal topics
        motivation_keywords = ["goal", "motivation", "achieve", "improve", "better", "progress"]
        if any(keyword in message_lower for keyword in motivation_keywords):
            return "supportive"
        
        return self.current_model
    
    def process_message(self, user_message: str) -> Dict:
        """
        Process user message with full safety and empathy pipeline
        
        Args:
            user_message: The user's input message
            
        Returns:
            Dict with response, safety status, and metadata
        """
        # Check for safety issues
        severity, is_flagged, reason = self.safety_filter.analyze_message(user_message)
        
        # Store in history
        self.history.add_message("user", user_message, safety_flag=reason if is_flagged else None)
        
        # Handle flagged content
        if is_flagged and severity in ["critical", "warning"]:
            if severity == "critical":
                self.safety_filter.flag_message(user_message, severity, reason)
            
            safe_response = self.safety_filter.get_safe_response(severity, reason)
            self.history.add_message("assistant", safe_response)
            
            return {
                "response": safe_response,
                "is_safe": False,
                "severity": severity,
                "reason": reason,
                "model_used": "safety_protocol",
                "from_history": False
            }
        
        # Select appropriate model
        selected_model = self.select_best_model(user_message)
        
        # Check if similar topic was discussed before
        similar_history = self._find_similar_topic(user_message)
        
        # Generate empathetic response
        response = self._generate_empathetic_response(
            user_message, 
            selected_model, 
            similar_history
        )
        
        # Store response in history
        self.history.add_message("assistant", response)
        
        return {
            "response": response,
            "is_safe": True,
            "model_used": selected_model,
            "model_info": self.MODELS[selected_model],
            "from_history": len(similar_history) > 0,
            "context_used": similar_history
        }
    
    def _find_similar_topic(self, message: str) -> List[Dict]:
        """Find similar topics from conversation history"""
        if not self.history.get_user_messages():
            return []
        
        # Simple similarity check - look for shared keywords
        message_words = set(message.lower().split())
        similar_messages = []
        
        for hist_message in self.history.get_recent_context(num_messages=20):
            if hist_message["role"] == "user" and hist_message["content"] != message:
                hist_words = set(hist_message["content"].lower().split())
                similarity = len(message_words & hist_words) / len(message_words | hist_words)
                if similarity > 0.3:  # If more than 30% similar
                    similar_messages.append(hist_message)
        
        return similar_messages[:3]  # Return top 3 similar messages
    
    def _generate_empathetic_response(self, message: str, model: str, 
                                     similar_history: List[Dict]) -> str:
        """
        Generate empathetic response using selected model
        
        Args:
            message: User message
            model: Selected model to use
            similar_history: Related messages from history
            
        Returns:
            Empathetic response
        """
        response = ""
        
        # Start with empathy
        response += random.choice(self.EMPATHY_STARTERS) + " "
        
        # Add context from history if available
        if similar_history:
            response += f"I remember you mentioned something similar before. "
        
        # Add model-specific response
        model_info = self.MODELS[model]
        response += self._model_specific_response(message, model_info)
        
        # End with supportive closing
        response += " " + self._get_supportive_closing()
        
        return response
    
    def _model_specific_response(self, message: str, model_info: Dict) -> str:
        """Generate model-specific response content"""
        message_lower = message.lower()
        
        if model_info["specialization"] == "emotional support and understanding":
            return (
                f"What you're feeling is valid and important. "
                f"These moments are part of being human. "
                f"I'm here to listen and support you through this. "
                f"Tell me more about what's on your mind?"
            )
        
        elif model_info["specialization"] == "actionable advice and solutions":
            return (
                f"Let me help you find practical solutions. "
                f"Based on what you've shared, here are some steps we could consider: "
                f"1) First, let's understand the core issue clearly. "
                f"2) Then we can explore options together. "
                f"What aspect would you like to focus on first?"
            )
        
        elif model_info["specialization"] == "deep thinking and complex topics":
            return (
                f"This is a nuanced and important topic worth thinking deeply about. "
                f"There are multiple perspectives here, and your viewpoint matters. "
                f"Let's explore this together thoughtfully. "
                f"What's your initial thoughts on this?"
            )
        
        elif model_info["specialization"] == "encouragement and motivation":
            return (
                f"You're taking positive steps by engaging with this. "
                f"Progress happens one moment at a time, and you're moving forward. "
                f"Let's focus on what you can control and build from there. "
                f"What's your vision for how things could improve?"
            )
        
        return "I'm here to help. Tell me more about what you need."
    
    def _get_supportive_closing(self) -> str:
        """Get supportive closing line"""
        closings = [
            "I'm here for you.",
            "Remember, you're not alone in this.",
            "Your well-being matters.",
            "Let's take this one step at a time.",
            "I believe in you.",
            "You've got this."
        ]
        return random.choice(closings)
    
    def get_conversation_summary(self) -> Dict:
        """Get summary of current conversation"""
        return self.history.get_session_summary()
    
    def get_safety_report(self) -> Dict:
        """Get safety incident report"""
        return self.safety_filter.generate_incident_report()
    
    def save_session(self, filepath: str) -> None:
        """Save current session to file"""
        self.history.save_to_file(filepath)
        print(f"Session saved to {filepath}")
    
    def load_session(self, filepath: str) -> None:
        """Load previous session from file"""
        self.history.load_from_file(filepath)
        print(f"Session loaded from {filepath}")
    
    def get_model_info(self, model_name: Optional[str] = None) -> Dict:
        """Get information about available models"""
        if model_name:
            return self.MODELS.get(model_name, {})
        return self.MODELS

File: javascript/src/EmpathyAI.js
/**
 * Empathetic AI Module
 * Implements AI with empathy, context awareness, and ethical responses
 * Supports multiple models for different scenarios
 */

const ConversationHistory = require('./ConversationHistory');
const SafetyFilter = require('./SafetyFilter');

class EmpathyAI {
  /**
   * Initialize EmpathyAI
   * @param {string} userName - Name of the user for personalization
   * @param {string} defaultModel - Default model to use
   */
  constructor(userName = 'User', defaultModel = 'empathetic') {
    this.userName = userName;
    this.currentModel = defaultModel;
    this.history = new ConversationHistory();
    this.safetyFilter = new SafetyFilter('medium');

    // Model profiles for different contexts
    this.MODELS = {
      empathetic: {
        name: 'Empathetic Model',
        specialization: 'emotional support and understanding',
        traits: ['compassionate', 'validating', 'patient', 'non-judgmental'],
        bestFor: 'crisis support, emotional topics, personal struggles'
      },
      practical: {
        name: 'Practical Model',
        specialization: 'actionable advice and solutions',
        traits: ['efficient', 'focused', 'solution-oriented', 'clear'],
        bestFor: 'problem-solving, advice, technical questions'
      },
      analytical: {
        name: 'Analytical Model',
        specialization: 'deep thinking and complex topics',
        traits: ['thoughtful', 'thorough', 'balanced', 'nuanced'],
        bestFor: 'philosophy, ethics, complex analysis'
      },
      supportive: {
        name: 'Supportive Model',
        specialization: 'encouragement and motivation',
        traits: ['positive', 'encouraging', 'motivating', 'uplifting'],
        bestFor: 'motivation, goal-setting, self-improvement'
      }
    };

    this.EMPATHY_STARTERS = [
      'I understand how you feel.',
      'That sounds really challenging.',
      'Your feelings make complete sense.',
      "I hear you, and what you're going through matters.",
      "That's a difficult situation, and I appreciate you sharing it.",
      'I can see why that would be frustrating.',
      'Thank you for sharing that with me.',
      'It takes courage to talk about this.',
      'I appreciate your honesty.',
      'You are not alone in feeling this way.',
      'Many people experience this, and it is okay to talk about it.',
      'I want to support you as best I can.'
    ];
    this._lastStarter = null;
    this._lastClosing = null;
  }

  /**
   * Select the best model based on message content
   * @param {string} message - User message to analyze
   * @returns {string} Selected model name
   */
  selectBestModel(message) {
    const messageLower = message.toLowerCase();

    // Detect crisis or emotional distress
    const emotionalKeywords = [
      'feel', 'sad', 'lonely', 'depressed', 'anxious',
      'scared', 'worried', 'hurt', 'pain', 'suffering'
    ];
    if (emotionalKeywords.some(keyword => messageLower.includes(keyword))) {
      return 'empathetic';
    }

    // Detect problem-solving requests
    const problemKeywords = ['how to', 'help me', 'advice', 'solution', 'fix', 'problem'];
    if (problemKeywords.some(keyword => messageLower.includes(keyword))) {
      return 'practical';
    }

    // Detect philosophical/complex topics
    const complexKeywords = ['why', 'meaning', 'ethics', 'philosophy', 'right or wrong', 'moral'];
    if (complexKeywords.some(keyword => messageLower.includes(keyword))) {
      return 'analytical';
    }

    // Detect motivation/goal topics
    const motivationKeywords = ['goal', 'motivation', 'achieve', 'improve', 'better', 'progress'];
    if (motivationKeywords.some(keyword => messageLower.includes(keyword))) {
      return 'supportive';
    }

    return this.currentModel;
  }

  /**
   * Process user message with full safety and empathy pipeline
   * @param {string} userMessage - The user's input message
   * @returns {object}
   */
  processMessage(userMessage) {
    // Check for "need to talk" messages
    const needToTalkPattern = /\b(i\s+need\s+to\s+talk|need\s+to\s+talk\s+to\s+you|i\s+need\s+someone\s+to\s+talk\s+to|can\s+i\s+talk\s+to\s+you)\b/i;
    if (needToTalkPattern.test(userMessage)) {
      this.history.addMessage('user', userMessage, null);
      const talkResponses = [
        "What can I help you with?",
        "I'm here to listen. What can I help you with?",
        "Of course. What can I help you with?",
        "I'm here for you. What can I help you with?"
      ];
      const response = talkResponses[Math.floor(Math.random() * talkResponses.length)];
      this.history.addMessage('assistant', response);
      
      return {
        response,
        isSafe: true,
        modelUsed: 'talk_request_response',
        fromHistory: false
      };
    }

    // Check for thank you messages
    const thankYouPattern = /\b(thank you|thanks|thank u|thx|appreciate it|grateful)\b/i;
    if (thankYouPattern.test(userMessage)) {
      this.history.addMessage('user', userMessage, null);
      const gratitudeResponses = [
        "Of course, I am grateful to be of help to you.",
        "You're very welcome! I'm grateful to be of help to you.",
        "I'm happy to help! It's my pleasure to be here for you.",
        "Of course! I'm grateful I can be of help to you."
      ];
      const response = gratitudeResponses[Math.floor(Math.random() * gratitudeResponses.length)];
      this.history.addMessage('assistant', response);
      
      return {
        response,
        isSafe: true,
        modelUsed: 'gratitude_response',
        fromHistory: false
      };
    }

    // Check for safety issues
    const [severity, isFlaged, reason] = this.safetyFilter.analyzeMessage(userMessage);

    // Store in history
    this.history.addMessage('user', userMessage, reason ? reason : null);

    // Handle flagged content
    if (isFlaged && ['critical', 'warning'].includes(severity)) {
      if (severity === 'critical') {
        this.safetyFilter.flagMessage(userMessage, severity, reason);
      }

      const safeResponse = this.safetyFilter.getSafeResponse(severity, reason);
      this.history.addMessage('assistant', safeResponse);

      return {
        response: safeResponse,
        isSafe: false,
        severity,
        reason,
        modelUsed: 'safety_protocol',
        fromHistory: false
      };
    }

    // Select appropriate model
    const selectedModel = this.selectBestModel(userMessage);

    // Check if similar topic was discussed before
    const similarHistory = this._findSimilarTopic(userMessage);

    // Generate empathetic response
    const response = this._generateEmpatheticResponse(
      userMessage,
      selectedModel,
      similarHistory
    );

    // Store response in history
    this.history.addMessage('assistant', response);

    return {
      response,
      isSafe: true,
      modelUsed: selectedModel,
      modelInfo: this.MODELS[selectedModel],
      fromHistory: similarHistory.length > 0,
      contextUsed: similarHistory
    };
  }

  /**
   * Find similar topics from conversation history
   * @param {string} message - Current message
   * @returns {Array}
   */
  _findSimilarTopic(message) {
    if (this.history.getUserMessages().length === 0) {
      return [];
    }

    const messageWords = new Set(message.toLowerCase().split(/\s+/));
    const similarMessages = [];

    const recentContext = this.history.getRecentContext(20);
    for (const histMessage of recentContext) {
      if (histMessage.role === 'user' && histMessage.content !== message) {
        const histWords = new Set(histMessage.content.toLowerCase().split(/\s+/));
        
        // Calculate Jaccard similarity
        const intersection = new Set([...messageWords].filter(x => histWords.has(x)));
        const union = new Set([...messageWords, ...histWords]);
        const similarity = intersection.size / union.size;
        
        if (similarity > 0.3) {
          similarMessages.push(histMessage);
        }
      }
    }

    return similarMessages.slice(0, 3);
  }

  /**
   * Generate empathetic response using selected model
   * @param {string} message - User message
   * @param {string} model - Selected model to use
   * @param {Array} similarHistory - Related messages from history
   * @returns {string}
   */
  _generateEmpatheticResponse(message, model, similarHistory) {
    let response = '';

    // Start with empathy, avoid repeating last starter
    let starterIndex;
    do {
      starterIndex = Math.floor(Math.random() * this.EMPATHY_STARTERS.length);
    } while (this.EMPATHY_STARTERS.length > 1 && this._lastStarter === starterIndex);
    response += this.EMPATHY_STARTERS[starterIndex] + ' ';
    this._lastStarter = starterIndex;

    // Add context from history if available
    if (similarHistory.length > 0) {
      response += 'I remember you mentioned something similar before. ';
    }

    // Add model-specific response
    const modelInfo = this.MODELS[model];
    response += this._modelSpecificResponse(message, modelInfo);

    // End with supportive closing, avoid repeating last closing
    const closing = this._getSupportiveClosing();
    response += ' ' + closing;
    this._lastClosing = closing;

    return response;
  }

  /**
   * Generate model-specific response content
   * @param {string} message - User message
   * @param {object} modelInfo - Model information
   * @returns {string}
   */
  _modelSpecificResponse(message, modelInfo) {
    const specialization = modelInfo.specialization;

    if (specialization === 'emotional support and understanding') {
      const options = [
        "What you're feeling is valid and important. These moments are part of being human. I'm here to listen and support you through this. Tell me more about what's on your mind?",
        "It's okay to feel this way. I'm here to talk and listen whenever you need.",
        "Your emotions matter, and it's good that you're expressing them. Would you like to share more?",
        "You deserve compassion and understanding. I'm here to talk if you need me."
      ];
      return options[Math.floor(Math.random() * options.length)];
    }
    if (specialization === 'actionable advice and solutions') {
      const options = [
        'Let me help you find practical solutions. Based on what you\'ve shared, here are some steps we could consider: 1) First, let\'s understand the core issue clearly. 2) Then we can explore options together. What aspect would you like to focus on first?',
        'We can break this down into manageable steps. What is the first thing you want to tackle?',
        'I can offer some advice or resources if you want. What would be most helpful for you right now?',
        'Let\'s work together to find a way forward. What outcome would you like to see?'
      ];
      return options[Math.floor(Math.random() * options.length)];
    }
    if (specialization === 'deep thinking and complex topics') {
      const options = [
        'This is a nuanced and important topic worth thinking deeply about. There are multiple perspectives here, and your viewpoint matters. Let\'s explore this together thoughtfully. What\'s your initial thoughts on this?',
        'That\'s a profound question. There are many ways to look at it. What are your thoughts so far?',
        'It\'s great to reflect on these kinds of questions. Would you like to share your perspective?',
        'There is no single answer, but exploring it can be meaningful. What draws you to this topic?'
      ];
      return options[Math.floor(Math.random() * options.length)];
    }
    if (specialization === 'encouragement and motivation') {
      const options = [
        "You're taking positive steps by engaging with this. Progress happens one moment at a time, and you\'re moving forward. Let\'s focus on what you can control and build from there. What\'s your vision for how things could improve?",
        "Every step you take matters. What is one thing you feel good about today?",
        "You have strengths that can help you move forward. What motivates you right now?",
        "Small progress is still progress. What\'s one goal you want to work toward?"
      ];
      return options[Math.floor(Math.random() * options.length)];
    }
    return 'I\'m here to help. Tell me more about what you need.';
  }

  /**
   * Get supportive closing line
   * @returns {string}
   */
  _getSupportiveClosing() {
    const closings = [
      "I'm here for you.",
      "Remember, you're not alone in this.",
      "Your well-being matters.",
      "Let's take this one step at a time.",
      'I believe in you.',
      "You've got this.",
      "If you need to talk more, I'm here.",
      "Take care of yourself, you matter.",
      "You are stronger than you think.",
      "I'm always here to listen."
    ];
    // Avoid repeating the last closing
    let closing;
    do {
      closing = closings[Math.floor(Math.random() * closings.length)];
    } while (closings.length > 1 && closing === this._lastClosing);
    return closing;
  }

  /**
   * Get summary of current conversation
   * @returns {object}
   */
  getConversationSummary() {
    return this.history.getSessionSummary();
  }

  /**
   * Get safety incident report
   * @returns {object}
   */
  getSafetyReport() {
    return this.safetyFilter.generateIncidentReport();
  }

  /**
   * Save current session to file
   * @param {string} filepath - Path to save to
   */
  saveSession(filepath) {
    this.history.saveToFile(filepath);
    console.log(`Session saved to ${filepath}`);
  }

  /**
   * Load previous session from file
   * @param {string} filepath - Path to load from
   */
  loadSession(filepath) {
    this.history.loadFromFile(filepath);
    console.log(`Session loaded from ${filepath}`);
  }

  /**
   * Get information about available models
   * @param {string|null} modelName - Specific model name or null for all
   * @returns {object}
   */
  getModelInfo(modelName = null) {
    if (modelName) {
      return this.MODELS[modelName] || {};
    }
    return this.MODELS;
  }
}

module.exports = EmpathyAI;
